packages/vitest/src/node/watch-filter.ts

```diff
@@ -0,0 +1,163 @@
+import readline from 'node:readline'
+import c from 'picocolors'
+import { stdout } from '../utils'
+
+type FilterFunc = (keyword: string) => Promise<string[]> // Define the FilterFunc type if not already defined
+const MAX_RESULT_COUNT = 10
+const SELECTION_MAX_INDEX = 7
+const ESC = '\u001B['
+
+export class WatchFilter {
+  private filterRL: readline.Interface
+  private currentKeyword: string | undefined = undefined
+  private message: string
+  private results: string[] = []
+  private selectionIndex = -1
+  private onKeyPress?: (str: string, key: any) => void
+
+  constructor(message: string) {
+    this.message = message
+    this.filterRL = readline.createInterface({ input: process.stdin, escapeCodeTimeout: 50 })
+    readline.emitKeypressEvents(process.stdin, this.filterRL)
+    if (process.stdin.isTTY)
+      process.stdin.setRawMode(true)
+  }
+
+  public async filter(initial: string, filterFunc: FilterFunc): Promise<string | undefined> {
+    stdout().write(`${ESC}?25h`)
```

Comment:
I've decided not to use`prompts` since it does not go well with our own implementation.